```{r}
devtools::load_all()
recipeUpdate(force=TRUE)
recipeSearch(c("reference", "genome"))
rcp <- recipeLoad("reference_genome")
```

```{r}
## ensembl homo_sapiens GRCh38 primary -- tested!! 
inputs(rcp)
rcp$fasta <- "http://ftp.ensembl.org/pub/release-104/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.chromosome.MT.fa.gz"
inputs(rcp)  ## FIXME: doesn't show input value...
## requirements(rcp)[[2]] <- requireDocker("refindex")

res <- getData(cwl = rcp,
               outdir = "../SharedData/referenceGenome",
               prefix = "Homo_sapiens.GRCh38.dna.MT", 
               ## docker =F,
               showLog=TRUE,
               cwlArgs = "--tmpdir-prefix tmp/"
               )

## docker2sif(req2$dockerFile, paste0(req2$dockerImageId, ".sif"), "--remote")
## rcp$fasta <- "http://ftp.ensembl.org/pub/release-104/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz"
## res <- getData(rcp, outdir = "/projects/rpci/shared/references/Homo_sapiens.GRCh38", showLog = TRUE,
##                   prefix = "Homo_sapiens.GRCh38.dna.primary_assembly",
##                   version = "release-104",
##                   notes = paste("ensembl", "Homo_sapiens.GRCh38",
##                                 "primary_assembly",
##                                 "fai", "dict"),
##                   docker = "singularity")

## 1kg hg38
rcp$fasta <- "https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/GRCh38_reference_genome/GRCh38_full_analysis_set_plus_decoy_hla.fa"

res <- getData(rcp,
               outdir = "../SharedData/referenceGenome",
               showLog = TRUE,
               prefix = "GRCh38_full_analysis_set_plus_decoy_hla",
               ## version = "",
               notes = c("1000genomes","Homo_sapiens", "GRCh38",
                         "decoy_hla"), 
               docker = FALSE)

## hs37d5  -- tested!!!
rcp$fasta <- "https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/phase2_reference_assembly_sequence/hs37d5.fa.gz"

res <- getData(rcp,
               outdir = "../SharedData/referenceGenome",
               showLog = TRUE,
               prefix = "hs37d5",
               ## version = "",
               notes = c("1000genomes", "Homo_sapiens", "GRCh37"),
               docker = FALSE)


## BPPARAM <- BatchtoolsParam(workers = 1,
##                            cluster = "slurm",
##                            template = "~/slurm_rpci.tmpl",
##                            resources = list(ncpus = 2,
##                                             jobname = "index",                         
##                                             walltime = 60*60*4,
##                                             memory = 16000),
##                            log = TRUE, logdir = ".", progressbar = TRUE)
## bplapply(1,
##          runBatch,
##          libs = c("Rcwl", "tools"),
##          fun = getData,
##          cwl = rcp,
##          outdir = "/projects/rpci/shared/references/hs37d5",
##          showLog = TRUE,
##          prefix = "hs37d5",
##          version = "",
##          notes = paste("1000genomes", "20110707", "Homo_sapiens", "GRCh37",
##                        "fai", "dict", "bwa_index", "bwa_v0.7.17-r1188"),
##          docker = FALSE,
##          BPPARAM = BPPARAM)

##
## GRCm39 -- tested!!!
rcp$fasta <- "https://ftp.ensembl.org/pub/release-104/fasta/mus_musculus/dna/Mus_musculus.GRCm39.dna.toplevel.fa.gz"
res <- getData(rcp,
               outdir = "../SharedData/referenceGenome",
               showLog = TRUE,
               prefix = "Mus_musculus.GRCm39.dna.toplevel",
               ## version = "release-104",
               notes = c("us_musculus", "GRCm39", "toplevel", "release-104"),
               docker = FALSE)
## library(BiocParallel)
## BPPARAM <- BatchtoolsParam(workers = 1,
##                            cluster = "slurm",
##                            template = "~/slurm_rpci.tmpl",
##                            resources = list(ncpus = 2,
##                                             jobname = "index",                         
##                                             walltime = 60*60*4,
##                                             memory = 16000),
##                            log = TRUE, logdir = ".", progressbar = TRUE)

```
