<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="ReUseData">
<title>ReUseData: Workflow-based Data Recipes for Management of Reusable and Reproducible Data Resources â€¢ ReUseData</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="ReUseData: Workflow-based Data Recipes for Management of Reusable and Reproducible Data Resources">
<meta property="og:description" content="ReUseData">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">ReUseData</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/ReUseData_data.html">ReUseData: Reusable and Reproducible Data Management</a>
    <a class="dropdown-item" href="../articles/ReUseData_quickStart.html">ReUseData: Reusable and Reproducible Data Management - quick start</a>
    <a class="dropdown-item" href="../articles/ReUseData_recipe.html">ReUseData: Workflow-based Data Recipes for Management of Reusable and Reproducible Data Resources</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rworkflow/ReUseData/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>ReUseData: Workflow-based Data Recipes for Management of Reusable and Reproducible Data Resources</h1>
                        <h4 data-toc-skip class="author">Qian Liu</h4>
            
            <h4 data-toc-skip class="date">2023-11-22</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rworkflow/ReUseData/blob/HEAD/vignettes/ReUseData_recipe.Rmd" class="external-link"><code>vignettes/ReUseData_recipe.Rmd</code></a></small>
      <div class="d-none name"><code>ReUseData_recipe.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The growth in the volume and complexity of genomic data resources
over the past few decades poses both opportunities and challenges for
data reuse. Presently, reuse of data often involves similar
preprocessing steps in different research projects. Lack of a
standardized annotation strategy can lead to difficult-to-find and even
duplicated datasets, resulting in substantial inefficiencies and wasted
computing resources, especially for research collaborations and
bioinformatics core facilities. Tools such as <code>GoGetData</code> and
<code>AnnotationHub</code> have been developed to mitigate common
problems in managing and accessing curated genomic datasets. However,
their use can be limited due to software requirements (e.g., Conda <a href="https://conda.io" class="external-link uri">https://conda.io</a>), forms of data
representation or scope of data resources.<br>
To respond to the FAIR (findability, accessibility, interoperability,
and reusability) data principles that are being widely adopted and
organizational requirements for Data Management Plans (DMPs), here, we
introduce <code>ReUseData</code>, an <em>R/Bioconductor</em> software
tool to provide a systematic and versatile approach for standardized and
reproducible data management. <code>ReUseData</code> facilitates
transformation of shell or other ad hoc scripts for data preprocessing
into workflow-based data recipes. Evaluation of data recipes generate
curated data files in their generic formats (e.g., VCF, bed) with full
annotations for subsequent reuse.</p>
<p>This package focuses on the management of genomic data resources and
uses classes and functions from existing <em>Bioconductor</em> packages.
So we think it should be a good fit for the <em>Bioconductor</em>.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<ol style="list-style-type: decimal">
<li>Install the package from <em>Bioconductor</em>.</li>
</ol>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"ReUseData"</span><span class="op">)</span></span></code></pre></div>
<p>Use the development version:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"ReUseData"</span>, version <span class="op">=</span> <span class="st">"devel"</span><span class="op">)</span></span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Load the package and other packages used in this vignette into the R
session.</li>
</ol>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">Rcwl</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/rworkflow/ReUseData" class="external-link">ReUseData</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="project-resources">Project resources<a class="anchor" aria-label="anchor" href="#project-resources"></a>
</h2>
<div class="section level3">
<h3 id="reusedata-recipe-landing-pages">
<code>ReUseData</code> recipe landing pages<a class="anchor" aria-label="anchor" href="#reusedata-recipe-landing-pages"></a>
</h3>
<p>The project website <a href="https://rcwl.org/dataRecipes/" class="external-link uri">https://rcwl.org/dataRecipes/</a> contains all prebuilt data
recipes for public data downloading and curation. They are available for
direct use with convenient webpage searching. Each data recipe has a
landing page including recipe description (inputs, outputs, etc.) and
user instructions. <strong>Make sure to check the instructions of
eligible input parameter values before recipe evaluation.</strong> These
prebuilt data recipes demonstrate the use of software and can be taken
as templates for users to create their own recipes for protected
datasets.</p>
<p>There are many other <em>R</em> resources available on this main
website <a href="https://rcwl.org/" class="external-link uri">https://rcwl.org/</a>,
including package vignettes for <code>Rcwl</code>
and<code>RcwlPipelines</code>, <code>Rcwl</code> tutorial e-book, case
studies of using <code>RcwlPipelines</code> in preprocessing single-cell
RNA-seq data, etc.</p>
</div>
</div>
<div class="section level2">
<h2 id="reusedata-recipe-scripts">
<code>ReUseData</code> recipe scripts<a class="anchor" aria-label="anchor" href="#reusedata-recipe-scripts"></a>
</h2>
<p>The prebuilt data recipe scripts are included in the package, and are
physically residing in a dedicated <a href="https://github.com/rworkflow/ReUseDataRecipes" class="external-link">GitHub
repository</a>, which demonstrates the recipe construction for different
situations. The most common case is that a data recipe can manage
multiple data resources with different input parameters (species,
versions, etc.). For example, the <code>gencode_transcripts</code>
recipe download from GENCODE, unzip and index the transcript fasta file
for human or mouse with different versions. A simple data downloading
(using <code>wget</code>) for a specific file can be written as a data
recipe without any input parameter. For example, the data recipe
<code>gcp_broad_gatk_hg38_1000G_omni2.5</code>) downloads the
<code>1000G_omni2.5.hg38.vcf.gz</code> and the <code>tbi</code> index
files from Google Cloud Platform bucket for Broad reference data GATK
hg38.</p>
<p>If the data curation gets more complicated, say, multiple
command-line tools are to be involved, and <code>conda</code> can be
used to install required packages, or some secondary files are to be
generated and collected, the raw way of building a
<code>ReUseData</code> recipe using <code>Rcwl</code> functions is
recommended, which gives more flexibility and power to accommodate
different situations. An example recipe is the
<code>reference_genome</code> which downloads, formats, and index
reference genome data using tools of <code>samtools</code>,
<code>picard</code> and <code>bwa</code>, and manages multiple secondary
files besides the main fasta file for later reuse.</p>
</div>
<div class="section level2">
<h2 id="reusedata-core-functions">
<code>ReUseData</code> core functions<a class="anchor" aria-label="anchor" href="#reusedata-core-functions"></a>
</h2>
<p>Here we show the usage of 4 core functions <code>recipeMake</code>,
<code>recipeUpdate</code>, <code>recipeSearch</code>,
<code>recipeLoad</code> for constructing, updating, searching and
loading <code>ReUseData</code> recipes in <em>R</em>.</p>
<div class="section level3">
<h3 id="recipe-construction-and-evaluation">Recipe construction and evaluation<a class="anchor" aria-label="anchor" href="#recipe-construction-and-evaluation"></a>
</h3>
<p>One can construct a data recipe from scratch or convert existing
shell scripts for data processing into data recipes, by specifying input
parameters, and output globbing patterns using <code>recipeMake</code>
function. Then the data recipe is represented in <em>R</em> as an S4
class <code>cwlProcess</code>. Upon assigning values to the input
parameters, the recipe is ready to be evaluated to generate data of
interest. Here are two examples:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">script</span> <span class="op">&lt;-</span> <span class="st">'</span></span>
<span><span class="st">input=$1</span></span>
<span><span class="st">outfile=$2</span></span>
<span><span class="st">echo "Print the input: $input" &gt; $outfile.txt</span></span>
<span><span class="st">'</span></span></code></pre></div>
<p>Equivalently, we can load the shell script directly:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">script</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"echo_out.sh"</span>, package <span class="op">=</span> <span class="st">"ReUseData"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rcp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/recipeMake.html">recipeMake</a></span><span class="op">(</span>shscript <span class="op">=</span> <span class="va">script</span>,</span>
<span>                  paramID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"input"</span>, <span class="st">"outfile"</span><span class="op">)</span>,</span>
<span>                  paramType <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"string"</span>, <span class="st">"string"</span><span class="op">)</span>,</span>
<span>                  outputID <span class="op">=</span> <span class="st">"echoout"</span>,</span>
<span>                  outputGlob <span class="op">=</span> <span class="st">"*.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/cwlProcess-methods.html" class="external-link">inputs</a></span><span class="op">(</span><span class="va">rcp</span><span class="op">)</span></span>
<span><span class="co">#&gt; inputs:</span></span>
<span><span class="co">#&gt;   input (string):  </span></span>
<span><span class="co">#&gt;   outfile (string):</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/cwlProcess-methods.html" class="external-link">outputs</a></span><span class="op">(</span><span class="va">rcp</span><span class="op">)</span></span>
<span><span class="co">#&gt; outputs:</span></span>
<span><span class="co">#&gt; echoout:</span></span>
<span><span class="co">#&gt;   type: File[]</span></span>
<span><span class="co">#&gt;   outputBinding:</span></span>
<span><span class="co">#&gt;     glob: '*.txt'</span></span></code></pre></div>
<p>Evaluation of the data recipes are internally submitted as CWL
workflow tasks, which requires the latest version of
<code>cwltool</code>. Here we have used <code>basilisk</code> to
initiate a conda environment and install the <code>cwltool</code> in
that environment if it is not available (or only older versions are
available) in the computer system.</p>
<p>We can install cwltool first to make sure a cwl-runner is
available.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/invisible.html" class="external-link">invisible</a></span><span class="op">(</span><span class="fu">Rcwl</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/install_cwltool.html" class="external-link">install_cwltool</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rcp</span><span class="op">$</span><span class="va">input</span> <span class="op">&lt;-</span> <span class="st">"Hello World!"</span></span>
<span><span class="va">rcp</span><span class="op">$</span><span class="va">outfile</span> <span class="op">&lt;-</span> <span class="st">"outfile"</span></span>
<span><span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"SharedData"</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getData.html">getData</a></span><span class="op">(</span><span class="va">rcp</span>,</span>
<span>               outdir <span class="op">=</span> <span class="va">outdir</span>,</span>
<span>               notes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"echo"</span>, <span class="st">"hello"</span>, <span class="st">"world"</span>, <span class="st">"txt"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; }<span style="color: #000000; font-weight: bold;">INFO</span> Final process status is success</span></span></code></pre></div>
<p>Letâ€™s take a look at the output file, which is successfully generated
in user-specified directory and grabbed through the
<code>outputGlob</code> argument. For more details of the
<code>getData</code> function for recipe evaluation, check the other
vignette for <a href="ReUseData_data.html">reusable data
management</a>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span><span class="op">$</span><span class="va">out</span></span>
<span><span class="co">#&gt; [1] "/tmp/RtmpGKcx5v/SharedData/outfile.txt"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readLines.html" class="external-link">readLines</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">out</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Print the input: Hello World!"</span></span></code></pre></div>
<p>Here we show a more complex example where the shell script has
required command line tools. When specific tools are needed for the data
processing, users just need to add their names in the
<code>requireTools</code> argument in <code>recipeMake</code> function,
and then add <code>conda = TRUE</code> when evaluating the recipe with
<code>getData</code> function. Then these tools will be automatically
installed by initiating a conda environment and the script can be
successfully run in that environment.</p>
<p>This function promotes data reproducibility across different
computing platforms, and removes barrier of using sophisticated
bioinformatics tools by less experienced users.</p>
<p>The following code chunk is not evaluated for time-limit of package
building but can be evaluated by users.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">shfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"gencode_transcripts.sh"</span>,</span>
<span>                      package <span class="op">=</span> <span class="st">"ReUseData"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readLines.html" class="external-link">readLines</a></span><span class="op">(</span><span class="va">shfile</span><span class="op">)</span></span>
<span><span class="va">rcp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/recipeMake.html">recipeMake</a></span><span class="op">(</span>shscript <span class="op">=</span> <span class="va">shfile</span>,</span>
<span>                  paramID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"species"</span>, <span class="st">"version"</span><span class="op">)</span>,</span>
<span>                  paramType <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"string"</span>, <span class="st">"string"</span><span class="op">)</span>,</span>
<span>                  outputID <span class="op">=</span> <span class="st">"transcripts"</span>, </span>
<span>                  outputGlob <span class="op">=</span> <span class="st">"*.transcripts.fa*"</span>,</span>
<span>                  requireTools <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"wget"</span>, <span class="st">"gzip"</span>, <span class="st">"samtools"</span><span class="op">)</span></span>
<span>                  <span class="op">)</span></span>
<span><span class="va">rcp</span><span class="op">$</span><span class="va">species</span> <span class="op">&lt;-</span> <span class="st">"human"</span></span>
<span><span class="va">rcp</span><span class="op">$</span><span class="va">version</span> <span class="op">&lt;-</span> <span class="st">"42"</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getData.html">getData</a></span><span class="op">(</span><span class="va">rcp</span>,</span>
<span>        outdir <span class="op">=</span> <span class="va">outdir</span>,</span>
<span>        notes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gencode"</span>, <span class="st">"transcripts"</span>, <span class="st">"human"</span>, <span class="st">"42"</span><span class="op">)</span>,</span>
<span>        conda <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">res</span><span class="op">$</span><span class="va">output</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="recipe-caching-and-updating">Recipe caching and updating<a class="anchor" aria-label="anchor" href="#recipe-caching-and-updating"></a>
</h3>
<p><code><a href="../reference/recipeUpdate.html">recipeUpdate()</a></code> creates a local cache for data recipes
that are saved in specified GitHub repository (if first time use), syncs
and updates data recipes from the GitHub repo to local caching system,
so any newly added recipes can be readily accessed and loaded directly
in <em>R</em>.</p>
<p><strong>NOTE:</strong></p>
<ul>
<li>The <code>cachePath</code> argument need to match between
<code>recipeUpdate</code>, <code>recipeLoad</code> and
<code>recipeSearch</code> functions.</li>
<li>use <code>force=TRUE</code> when any old recipes that are previously
cached are updated.</li>
<li>use <code>remote = TRUE</code>to sync with remote GitHub
repositories. By default, it syncs with <code>ReUseDataRecipe</code>
GitHub repository](<a href="https://github.com/rworkflow/ReUseDataRecipe" class="external-link uri">https://github.com/rworkflow/ReUseDataRecipe</a>) for
public, prebuilt data recipes. <code>repo</code> can also be a private
GitHub repository.</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## First time use</span></span>
<span><span class="fu"><a href="../reference/recipeUpdate.html">recipeUpdate</a></span><span class="op">(</span>cachePath <span class="op">=</span> <span class="st">"ReUseDataRecipe"</span>,</span>
<span>             force <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; NOTE: existing caches will be removed and regenerated!</span></span>
<span><span class="co">#&gt; Updating recipes...</span></span>
<span><span class="co">#&gt; bowtie2_index.R added</span></span>
<span><span class="co">#&gt; echo_out.R added</span></span>
<span><span class="co">#&gt; ensembl_liftover.R added</span></span>
<span><span class="co">#&gt; gcp_broad_gatk_hg19.R added</span></span>
<span><span class="co">#&gt; gcp_broad_gatk_hg38.R added</span></span>
<span><span class="co">#&gt; gcp_gatk_mutect2_b37.R added</span></span>
<span><span class="co">#&gt; gcp_gatk_mutect2_hg38.R added</span></span>
<span><span class="co">#&gt; gencode_annotation.R added</span></span>
<span><span class="co">#&gt; gencode_genome_grch38.R added</span></span>
<span><span class="co">#&gt; gencode_transcripts.R added</span></span>
<span><span class="co">#&gt; hisat2_index.R added</span></span>
<span><span class="co">#&gt; reference_genome.R added</span></span>
<span><span class="co">#&gt; salmon_index.R added</span></span>
<span><span class="co">#&gt; STAR_index.R added</span></span>
<span><span class="co">#&gt; ucsc_database.R added</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; recipeHub with 15 records</span></span>
<span><span class="co">#&gt; cache path:  /home/qhu/.cache/R/ReUseDataRecipe </span></span>
<span><span class="co">#&gt; # recipeSearch() to query specific recipes using multipe keywords</span></span>
<span><span class="co">#&gt; # recipeUpdate() to update the local recipe cache</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;            name               </span></span>
<span><span class="co">#&gt;   BFC496 | bowtie2_index      </span></span>
<span><span class="co">#&gt;   BFC497 | echo_out           </span></span>
<span><span class="co">#&gt;   BFC498 | ensembl_liftover   </span></span>
<span><span class="co">#&gt;   BFC499 | gcp_broad_gatk_hg19</span></span>
<span><span class="co">#&gt;   BFC500 | gcp_broad_gatk_hg38</span></span>
<span><span class="co">#&gt;   ...      ...                </span></span>
<span><span class="co">#&gt;   BFC506 | hisat2_index       </span></span>
<span><span class="co">#&gt;   BFC507 | reference_genome   </span></span>
<span><span class="co">#&gt;   BFC508 | salmon_index       </span></span>
<span><span class="co">#&gt;   BFC509 | STAR_index         </span></span>
<span><span class="co">#&gt;   BFC510 | ucsc_database</span></span></code></pre></div>
<p>To sync the local recipe cache with remote GitHub repository.
Currently the remote data recipes on GitHub are the same as the recipes
in package (so not evaluted here to avoid duplicate messages). We will
do our best to keep current of the data recipes in package development
version with the remote GitHub repository.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/recipeUpdate.html">recipeUpdate</a></span><span class="op">(</span>remote <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>             repos <span class="op">=</span> <span class="st">"rworkflow/ReUseDataRecipe"</span><span class="op">)</span>  <span class="co">## can be private repo</span></span></code></pre></div>
<p><code>recipeUpdate</code> returns a <code>recipeHub</code> object
with a list of all available recipes. One can subset the list with
<code>[</code> and use getter functions <code><a href="../reference/recipeHub-class.html">recipeNames()</a></code> to
get the recipe names which can then be passed to the
<code><a href="../reference/recipeSearch.html">recipeSearch()</a></code> or <code><a href="../reference/recipeLoad.html">recipeLoad()</a></code>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rh</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/recipeUpdate.html">recipeUpdate</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Updating recipes...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/methods/is.html" class="external-link">is</a></span><span class="op">(</span><span class="va">rh</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "recipeHub"             "cwlHub"                "BiocFileCacheReadOnly"</span></span>
<span><span class="co">#&gt; [4] "BiocFileCacheBase"</span></span>
<span><span class="va">rh</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="co">#&gt; recipeHub with 1 records</span></span>
<span><span class="co">#&gt; cache path:  /home/qhu/.cache/R/ReUseDataRecipe </span></span>
<span><span class="co">#&gt; # recipeSearch() to query specific recipes using multipe keywords</span></span>
<span><span class="co">#&gt; # recipeUpdate() to update the local recipe cache</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;            name         </span></span>
<span><span class="co">#&gt;   BFC496 | bowtie2_index</span></span>
<span><span class="fu"><a href="../reference/recipeHub-class.html">recipeNames</a></span><span class="op">(</span><span class="va">rh</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "bowtie2_index"         "echo_out"              "ensembl_liftover"     </span></span>
<span><span class="co">#&gt;  [4] "gcp_broad_gatk_hg19"   "gcp_broad_gatk_hg38"   "gcp_gatk_mutect2_b37" </span></span>
<span><span class="co">#&gt;  [7] "gcp_gatk_mutect2_hg38" "gencode_annotation"    "gencode_genome_grch38"</span></span>
<span><span class="co">#&gt; [10] "gencode_transcripts"   "hisat2_index"          "reference_genome"     </span></span>
<span><span class="co">#&gt; [13] "salmon_index"          "STAR_index"            "ucsc_database"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="recipe-searching-and-loading">Recipe searching and loading<a class="anchor" aria-label="anchor" href="#recipe-searching-and-loading"></a>
</h3>
<p>Cached data recipes can be searched using multiple keywords to match
the recipe name. It returns a <code>recipeHub</code> object with a list
of recipes available.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/recipeSearch.html">recipeSearch</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; recipeHub with 15 records</span></span>
<span><span class="co">#&gt; cache path:  /home/qhu/.cache/R/ReUseDataRecipe </span></span>
<span><span class="co">#&gt; # recipeSearch() to query specific recipes using multipe keywords</span></span>
<span><span class="co">#&gt; # recipeUpdate() to update the local recipe cache</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;            name               </span></span>
<span><span class="co">#&gt;   BFC496 | bowtie2_index      </span></span>
<span><span class="co">#&gt;   BFC497 | echo_out           </span></span>
<span><span class="co">#&gt;   BFC498 | ensembl_liftover   </span></span>
<span><span class="co">#&gt;   BFC499 | gcp_broad_gatk_hg19</span></span>
<span><span class="co">#&gt;   BFC500 | gcp_broad_gatk_hg38</span></span>
<span><span class="co">#&gt;   ...      ...                </span></span>
<span><span class="co">#&gt;   BFC506 | hisat2_index       </span></span>
<span><span class="co">#&gt;   BFC507 | reference_genome   </span></span>
<span><span class="co">#&gt;   BFC508 | salmon_index       </span></span>
<span><span class="co">#&gt;   BFC509 | STAR_index         </span></span>
<span><span class="co">#&gt;   BFC510 | ucsc_database</span></span>
<span><span class="fu"><a href="../reference/recipeSearch.html">recipeSearch</a></span><span class="op">(</span><span class="st">"gencode"</span><span class="op">)</span></span>
<span><span class="co">#&gt; recipeHub with 3 records</span></span>
<span><span class="co">#&gt; cache path:  /home/qhu/.cache/R/ReUseDataRecipe </span></span>
<span><span class="co">#&gt; # recipeSearch() to query specific recipes using multipe keywords</span></span>
<span><span class="co">#&gt; # recipeUpdate() to update the local recipe cache</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;            name                 </span></span>
<span><span class="co">#&gt;   BFC503 | gencode_annotation   </span></span>
<span><span class="co">#&gt;   BFC504 | gencode_genome_grch38</span></span>
<span><span class="co">#&gt;   BFC505 | gencode_transcripts</span></span>
<span><span class="fu"><a href="../reference/recipeSearch.html">recipeSearch</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"STAR"</span>, <span class="st">"index"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; recipeHub with 1 records</span></span>
<span><span class="co">#&gt; cache path:  /home/qhu/.cache/R/ReUseDataRecipe </span></span>
<span><span class="co">#&gt; # recipeSearch() to query specific recipes using multipe keywords</span></span>
<span><span class="co">#&gt; # recipeUpdate() to update the local recipe cache</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;            name      </span></span>
<span><span class="co">#&gt;   BFC509 | STAR_index</span></span></code></pre></div>
<p>Recipes can be directly loaded into <em>R</em> using
<code>recipeLoad</code> function with user assigned name or the original
recipe name. Once the recipe is successfully loaded, a message will be
returned with recipe instructions.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rcp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/recipeLoad.html">recipeLoad</a></span><span class="op">(</span><span class="st">"STAR_index"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Note: you need to assign a name for the recipe: rcpName &lt;- recipeLoad('xx')</span></span>
<span><span class="co">#&gt; Data recipe loaded!</span></span>
<span><span class="co">#&gt; Use inputs() to check required input parameters before evaluation.</span></span>
<span><span class="co">#&gt; Check here: https://rcwl.org/dataRecipes/STAR_index.html</span></span>
<span><span class="co">#&gt; for user instructions (e.g., eligible input values, data source, etc.)</span></span></code></pre></div>
<p><strong>NOTE</strong> Use <code>return=FALSE</code> if you want to
keep the original recipe name, or if multiple recipes are to be
loaded.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/recipeLoad.html">recipeLoad</a></span><span class="op">(</span><span class="st">"STAR_index"</span>, return <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Data recipe loaded!</span></span>
<span><span class="co">#&gt; Use inputs(STAR_index) to check required input parameters before evaluation.</span></span>
<span><span class="co">#&gt; Check here: https://rcwl.org/dataRecipes/STAR_index.html</span></span>
<span><span class="co">#&gt; for user instructions (e.g., eligible input values, data source, etc.)</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="va">rcp</span>, <span class="va">STAR_index</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/recipeLoad.html">recipeLoad</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ensembl_liftover"</span>, <span class="st">"gencode_annotation"</span><span class="op">)</span>, return<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Data recipe loaded!</span></span>
<span><span class="co">#&gt; Use inputs(ensembl_liftover) to check required input parameters before evaluation.</span></span>
<span><span class="co">#&gt; Check here: https://rcwl.org/dataRecipes/ensembl_liftover.html</span></span>
<span><span class="co">#&gt; for user instructions (e.g., eligible input values, data source, etc.)</span></span>
<span><span class="co">#&gt; Data recipe loaded!</span></span>
<span><span class="co">#&gt; Use inputs(gencode_annotation) to check required input parameters before evaluation.</span></span>
<span><span class="co">#&gt; Check here: https://rcwl.org/dataRecipes/gencode_annotation.html</span></span>
<span><span class="co">#&gt; for user instructions (e.g., eligible input values, data source, etc.)</span></span></code></pre></div>
<p>Itâ€™s important to check the required <code><a href="https://rdrr.io/pkg/Rcwl/man/cwlProcess-methods.html" class="external-link">inputs()</a></code> of the
recipe and the recipe landing page for eligible input parameter values
before evaluating the recipe to generate data of interest.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/cwlProcess-methods.html" class="external-link">inputs</a></span><span class="op">(</span><span class="va">STAR_index</span><span class="op">)</span></span>
<span><span class="co">#&gt; inputs:</span></span>
<span><span class="co">#&gt;   ref (reference genome)   ( string|File ): </span></span>
<span><span class="co">#&gt;   gtf (GTF)   ( string|File ): </span></span>
<span><span class="co">#&gt;   genomeDir (genomeDir)  (string):  </span></span>
<span><span class="co">#&gt;   threads (threads)  (int):  </span></span>
<span><span class="co">#&gt;   sjdb (sjdbOverhang)  (int):  100</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/cwlProcess-methods.html" class="external-link">inputs</a></span><span class="op">(</span><span class="va">ensembl_liftover</span><span class="op">)</span></span>
<span><span class="co">#&gt; inputs:</span></span>
<span><span class="co">#&gt;   species (species)  (string):  </span></span>
<span><span class="co">#&gt;   from (from)  (string):  </span></span>
<span><span class="co">#&gt;   to (to)  (string):</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/cwlProcess-methods.html" class="external-link">inputs</a></span><span class="op">(</span><span class="va">gencode_annotation</span><span class="op">)</span></span>
<span><span class="co">#&gt; inputs:</span></span>
<span><span class="co">#&gt;   species (species)  (string):  </span></span>
<span><span class="co">#&gt;   version (version)  (string):</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="sessioninfo">SessionInfo<a class="anchor" aria-label="anchor" href="#sessioninfo"></a>
</h2>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">#&gt; Platform: x86_64-conda-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS/LAPACK: /home/qhu/miniconda3/envs/rcwl/lib/libopenblasp-r0.3.23.so;  LAPACK version 3.11.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: America/New_York</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] ReUseData_1.3.2     Rcwl_1.18.0         S4Vectors_0.40.1   </span></span>
<span><span class="co">#&gt; [4] BiocGenerics_0.48.1 yaml_2.3.7          BiocStyle_2.30.0   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] tidyselect_1.2.0      dplyr_1.1.4           blob_1.2.4           </span></span>
<span><span class="co">#&gt;  [4] filelock_1.0.2        R.utils_2.12.3        fastmap_1.1.1        </span></span>
<span><span class="co">#&gt;  [7] BiocFileCache_2.10.1  promises_1.2.1        digest_0.6.33        </span></span>
<span><span class="co">#&gt; [10] base64url_1.4         mime_0.12             lifecycle_1.0.4      </span></span>
<span><span class="co">#&gt; [13] ellipsis_0.3.2        RSQLite_2.3.3         magrittr_2.0.3       </span></span>
<span><span class="co">#&gt; [16] compiler_4.3.1        rlang_1.1.2           sass_0.4.7           </span></span>
<span><span class="co">#&gt; [19] progress_1.2.2        tools_4.3.1           utf8_1.2.4           </span></span>
<span><span class="co">#&gt; [22] data.table_1.14.8     knitr_1.45            prettyunits_1.2.0    </span></span>
<span><span class="co">#&gt; [25] brew_1.0-8            htmlwidgets_1.6.2     curl_5.1.0           </span></span>
<span><span class="co">#&gt; [28] bit_4.0.5             reticulate_1.34.0     RColorBrewer_1.1-3   </span></span>
<span><span class="co">#&gt; [31] batchtools_0.9.17     BiocParallel_1.36.0   withr_2.5.2          </span></span>
<span><span class="co">#&gt; [34] purrr_1.0.2           desc_1.4.2            R.oo_1.25.0          </span></span>
<span><span class="co">#&gt; [37] grid_4.3.1            fansi_1.0.5           git2r_0.32.0         </span></span>
<span><span class="co">#&gt; [40] xtable_1.8-4          cli_3.6.1             rmarkdown_2.25       </span></span>
<span><span class="co">#&gt; [43] DiagrammeR_1.0.10     crayon_1.5.2          ragg_1.2.5           </span></span>
<span><span class="co">#&gt; [46] generics_0.1.3        httr_1.4.7            visNetwork_2.1.2     </span></span>
<span><span class="co">#&gt; [49] DBI_1.1.3             cachem_1.0.8          stringr_1.5.1        </span></span>
<span><span class="co">#&gt; [52] parallel_4.3.1        BiocManager_1.30.22   basilisk_1.14.1      </span></span>
<span><span class="co">#&gt; [55] vctrs_0.6.4           Matrix_1.6-3          jsonlite_1.8.7       </span></span>
<span><span class="co">#&gt; [58] dir.expiry_1.10.0     bookdown_0.36         hms_1.1.3            </span></span>
<span><span class="co">#&gt; [61] bit64_4.0.5           systemfonts_1.0.4     jquerylib_0.1.4      </span></span>
<span><span class="co">#&gt; [64] RcwlPipelines_1.18.0  glue_1.6.2            pkgdown_2.0.7        </span></span>
<span><span class="co">#&gt; [67] codetools_0.2-19      stringi_1.8.1         later_1.3.1          </span></span>
<span><span class="co">#&gt; [70] tibble_3.2.1          pillar_1.9.0          basilisk.utils_1.14.1</span></span>
<span><span class="co">#&gt; [73] rappdirs_0.3.3        htmltools_0.5.7       R6_2.5.1             </span></span>
<span><span class="co">#&gt; [76] dbplyr_2.4.0          textshaping_0.3.6     rprojroot_2.0.4      </span></span>
<span><span class="co">#&gt; [79] evaluate_0.23         shiny_1.8.0           lattice_0.22-5       </span></span>
<span><span class="co">#&gt; [82] R.methodsS3_1.8.2     png_0.1-8             backports_1.4.1      </span></span>
<span><span class="co">#&gt; [85] memoise_2.0.1         httpuv_1.6.12         bslib_0.5.1          </span></span>
<span><span class="co">#&gt; [88] Rcpp_1.0.11           checkmate_2.3.0       xfun_0.41            </span></span>
<span><span class="co">#&gt; [91] fs_1.6.3              pkgconfig_2.0.3</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Qian Liu.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
