<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="ReUseData">
<title>ReUseData: Reusable and Reproducible Data Management • ReUseData</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="ReUseData: Reusable and Reproducible Data Management">
<meta property="og:description" content="ReUseData">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">ReUseData</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/ReUseData_data.html">ReUseData: Reusable and Reproducible Data Management</a>
    <a class="dropdown-item" href="../articles/ReUseData_quickStart.html">ReUseData: Reusable and Reproducible Data Management - quick start</a>
    <a class="dropdown-item" href="../articles/ReUseData_recipe.html">ReUseData: Workflow-based Data Recipes for Management of Reusable and Reproducible Data Resources</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rworkflow/ReUseData/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>ReUseData: Reusable and Reproducible Data Management</h1>
                        <h4 data-toc-skip class="author">Qian Liu, Qiang
Hu, Song Liu, Alan Hutson, Martin Morgan</h4>
            <address class="author_afil">
      Roswell Park Comprehensive Cancer Center<br><h4 data-toc-skip class="date">2023-11-22</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rworkflow/ReUseData/blob/HEAD/vignettes/ReUseData_data.Rmd" class="external-link"><code>vignettes/ReUseData_data.Rmd</code></a></small>
      <div class="d-none name"><code>ReUseData_data.Rmd</code></div>
    </address>
</div>

    
    
<p><code>ReUseData</code> provides functionalities to construct
workflow-based data recipes for fully tracked and reproducible data
processing. Evaluation of data recipes generates curated data resources
in their generic formats (e.g., VCF, bed), as well as a YAML manifest
file recording the recipe parameters, data annotations, and data file
paths for subsequent reuse. The datasets are locally cached using a
database infrastructure, where updating and searching of specific data
is made easy.</p>
<p>The data reusability is assured through cloud hosting and enhanced
interoperability with downstream software tools or analysis workflows.
The workflow strategy enables cross platform reproducibility of curated
data resources.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<ol style="list-style-type: decimal">
<li>Install the package from <em>Bioconductor</em>.</li>
</ol>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"ReUseData"</span><span class="op">)</span></span></code></pre></div>
<p>Use the development version:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"ReUseData"</span>, version <span class="op">=</span> <span class="st">"devel"</span><span class="op">)</span></span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Load the package and other packages used in this vignette into the R
session.</li>
</ol>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">Rcwl</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/rworkflow/ReUseData" class="external-link">ReUseData</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="reusedata-core-functions-for-data-management">
<code>ReUseData</code> core functions for data management<a class="anchor" aria-label="anchor" href="#reusedata-core-functions-for-data-management"></a>
</h2>
<p>Here we introduce the core functions of <code>ReUseData</code> for
data management and reuse: <code>getData</code> for reproducible data
generation, <code>dataUpdate</code> for syncing and updating data cache,
and <code>dataSearch</code> for multi-keywords searching of dataset of
interest.</p>
<div class="section level3">
<h3 id="data-generation">Data generation<a class="anchor" aria-label="anchor" href="#data-generation"></a>
</h3>
<p>First, we can construct data recipes by transforming shell or other
ad hoc data preprocessing scripts into workflow-based data recipes. Some
prebuilt data recipes for public data resources (e.g., downloading,
unzipping and indexing) are available for direct use through
<code>recipeSearch</code> and <code>recipeLoad</code> functions. Then we
will assign values to the input parameters and evaluate the recipe to
generate data of interest.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## set cache in tempdir for test</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Sys.setenv.html" class="external-link">Sys.setenv</a></span><span class="op">(</span>cachePath <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"cache"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/recipeUpdate.html">recipeUpdate</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Updating recipes...</span></span>
<span><span class="co">#&gt; bowtie2_index.R added</span></span>
<span><span class="co">#&gt; echo_out.R added</span></span>
<span><span class="co">#&gt; ensembl_liftover.R added</span></span>
<span><span class="co">#&gt; gcp_broad_gatk_hg19.R added</span></span>
<span><span class="co">#&gt; gcp_broad_gatk_hg38.R added</span></span>
<span><span class="co">#&gt; gcp_gatk_mutect2_b37.R added</span></span>
<span><span class="co">#&gt; gcp_gatk_mutect2_hg38.R added</span></span>
<span><span class="co">#&gt; gencode_annotation.R added</span></span>
<span><span class="co">#&gt; gencode_genome_grch38.R added</span></span>
<span><span class="co">#&gt; gencode_transcripts.R added</span></span>
<span><span class="co">#&gt; hisat2_index.R added</span></span>
<span><span class="co">#&gt; reference_genome.R added</span></span>
<span><span class="co">#&gt; salmon_index.R added</span></span>
<span><span class="co">#&gt; STAR_index.R added</span></span>
<span><span class="co">#&gt; ucsc_database.R added</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; recipeHub with 15 records</span></span>
<span><span class="co">#&gt; cache path:  /tmp/RtmpXMx6sd/cache/ReUseDataRecipe </span></span>
<span><span class="co">#&gt; # recipeSearch() to query specific recipes using multipe keywords</span></span>
<span><span class="co">#&gt; # recipeUpdate() to update the local recipe cache</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;           name               </span></span>
<span><span class="co">#&gt;   BFC1  | bowtie2_index      </span></span>
<span><span class="co">#&gt;   BFC2  | echo_out           </span></span>
<span><span class="co">#&gt;   BFC3  | ensembl_liftover   </span></span>
<span><span class="co">#&gt;   BFC4  | gcp_broad_gatk_hg19</span></span>
<span><span class="co">#&gt;   BFC5  | gcp_broad_gatk_hg38</span></span>
<span><span class="co">#&gt;   ...     ...                </span></span>
<span><span class="co">#&gt;   BFC11 | hisat2_index       </span></span>
<span><span class="co">#&gt;   BFC12 | reference_genome   </span></span>
<span><span class="co">#&gt;   BFC13 | salmon_index       </span></span>
<span><span class="co">#&gt;   BFC14 | STAR_index         </span></span>
<span><span class="co">#&gt;   BFC15 | ucsc_database</span></span>
<span><span class="fu"><a href="../reference/recipeSearch.html">recipeSearch</a></span><span class="op">(</span><span class="st">"echo"</span><span class="op">)</span></span>
<span><span class="co">#&gt; recipeHub with 1 records</span></span>
<span><span class="co">#&gt; cache path:  /tmp/RtmpXMx6sd/cache/ReUseDataRecipe </span></span>
<span><span class="co">#&gt; # recipeSearch() to query specific recipes using multipe keywords</span></span>
<span><span class="co">#&gt; # recipeUpdate() to update the local recipe cache</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;          name    </span></span>
<span><span class="co">#&gt;   BFC2 | echo_out</span></span>
<span><span class="va">echo_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/recipeLoad.html">recipeLoad</a></span><span class="op">(</span><span class="st">"echo_out"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Note: you need to assign a name for the recipe: rcpName &lt;- recipeLoad('xx')</span></span>
<span><span class="co">#&gt; Data recipe loaded!</span></span>
<span><span class="co">#&gt; Use inputs() to check required input parameters before evaluation.</span></span>
<span><span class="co">#&gt; Check here: https://rcwl.org/dataRecipes/echo_out.html</span></span>
<span><span class="co">#&gt; for user instructions (e.g., eligible input values, data source, etc.)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/cwlProcess-methods.html" class="external-link">inputs</a></span><span class="op">(</span><span class="va">echo_out</span><span class="op">)</span></span>
<span><span class="co">#&gt; inputs:</span></span>
<span><span class="co">#&gt;   input (input)  (string):  </span></span>
<span><span class="co">#&gt;   outfile (outfile)  (string):</span></span></code></pre></div>
<p>Users can then assign values to the input parameters, and evaluate
the recipe (<code>getData</code>) to generate data of interest. Users
need to specify an output directory for all files (desired data file,
intermediate files that are internally generated as workflow scripts or
annotation files). Detailed notes for the data is encouraged which will
be used for keywords matching for later data search.</p>
<p>We can install cwltool first to make sure a cwl-runner is
available.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/invisible.html" class="external-link">invisible</a></span><span class="op">(</span><span class="fu">Rcwl</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/install_cwltool.html" class="external-link">install_cwltool</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">echo_out</span><span class="op">$</span><span class="va">input</span> <span class="op">&lt;-</span> <span class="st">"Hello World!"</span></span>
<span><span class="va">echo_out</span><span class="op">$</span><span class="va">outfile</span> <span class="op">&lt;-</span> <span class="st">"outfile"</span></span>
<span><span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"SharedData"</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getData.html">getData</a></span><span class="op">(</span><span class="va">echo_out</span>,</span>
<span>               outdir <span class="op">=</span> <span class="va">outdir</span>,</span>
<span>               notes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"echo"</span>, <span class="st">"hello"</span>, <span class="st">"world"</span>, <span class="st">"txt"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; }<span style="color: #000000; font-weight: bold;">INFO</span> Final process status is success</span></span></code></pre></div>
<p>The file path to newly generated dataset can be easily retrieved. It
can also be retrieved using <code><a href="../reference/dataSearch.html">dataSearch()</a></code> functions with
multiple keywords. Before that, <code><a href="../reference/dataUpdate.html">dataUpdate()</a></code> needs to be
done.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span><span class="op">$</span><span class="va">output</span></span>
<span><span class="co">#&gt; [1] "/tmp/RtmpXMx6sd/SharedData/outfile.txt"</span></span></code></pre></div>
<p>There are some automatically generated files to help track the data
recipe evaluation, including <code>*.sh</code> to record the original
shell script, <code>*.cwl</code> file as the official workflow script
which was internally submitted for data recipe evaluation,
<code>*.yml</code> file as part of CWL workflow evaluation, which also
record data annotations, and <code>*.md5</code> checksum file to
check/verify the integrity of generated data file.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">outdir</span>, pattern <span class="op">=</span> <span class="st">"echo"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "echo_out_Hello_World!_outfile.cwl" "echo_out_Hello_World!_outfile.md5"</span></span>
<span><span class="co">#&gt; [3] "echo_out_Hello_World!_outfile.sh"  "echo_out_Hello_World!_outfile.yml"</span></span></code></pre></div>
<p>The <code>*.yml</code> file contains information about recipe input
parameters, the file path to output file, the notes for the dataset, and
auto-added date for data generation time. A later data search using
<code><a href="../reference/dataSearch.html">dataSearch()</a></code> will refer to this file for keywords
match.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/readLines.html" class="external-link">readLines</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">yml</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "input: Hello World!"                             </span></span>
<span><span class="co">#&gt; [2] "outfile: outfile"                                </span></span>
<span><span class="co">#&gt; [3] "# output: /tmp/RtmpXMx6sd/SharedData/outfile.txt"</span></span>
<span><span class="co">#&gt; [4] "# notes: echo hello world txt"                   </span></span>
<span><span class="co">#&gt; [5] "# date: 2023-11-22 22:21:02.021824"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="data-caching-updating-and-searching">Data caching, updating and searching<a class="anchor" aria-label="anchor" href="#data-caching-updating-and-searching"></a>
</h3>
<p><code><a href="../reference/dataUpdate.html">dataUpdate()</a></code> creates (if first time use), syncs and
updates the local cache for curated datasets. It finds and reads all the
<code>.yml</code> files recursively in the provided data folder, creates
a cache record for each dataset that is associated (including newly
generated ones with <code><a href="../reference/getData.html">getData()</a></code>), and updates the local cache
for later data searching and reuse.</p>
<p><strong>IMPORTANT:</strong> It is recommended that users create a
specified folder for data archival (e.g.,
<code>file/path/to/SharedData</code>) that other group members have
access to, and use sub-folders for different kinds of datasets (e.g.,
those generated from same recipe).</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">dh</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dataUpdate.html">dataUpdate</a></span><span class="op">(</span>dir <span class="op">=</span> <span class="va">outdir</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Updating data record...</span></span>
<span><span class="co">#&gt; outfile.txt added</span></span>
<span><span class="co">#&gt; dataHub with 1 records</span></span>
<span><span class="co">#&gt; cache path:  /tmp/RtmpXMx6sd/cache/ReUseData </span></span>
<span><span class="co">#&gt; # dataUpdate() to update the local data cache</span></span>
<span><span class="co">#&gt; # dataSearch() to query a specific dataset</span></span>
<span><span class="co">#&gt; # Additional information can be retrieved using: </span></span>
<span><span class="co">#&gt; # dataNames(), dataParams(), dataNotes(), dataPaths(), dataTag() or mcols()</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;          name        Path                                  </span></span>
<span><span class="co">#&gt;   BFC1 | outfile.txt /tmp/RtmpXMx6sd/SharedData/outfile.txt</span></span></code></pre></div>
<p><code>dataUpdate</code> and <code>dataSearch</code> return a
<code>dataHub</code> object with a list of all available or matching
datasets.</p>
<p>One can subset the list with <code>[</code> and use getter functions
to retrieve the annotation information about the data, e.g., data names,
parameters values to the recipe, notes, tags, and the corresponding yaml
file.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dh</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="co">#&gt; dataHub with 1 records</span></span>
<span><span class="co">#&gt; cache path:  /tmp/RtmpXMx6sd/cache/ReUseData </span></span>
<span><span class="co">#&gt; # dataUpdate() to update the local data cache</span></span>
<span><span class="co">#&gt; # dataSearch() to query a specific dataset</span></span>
<span><span class="co">#&gt; # Additional information can be retrieved using: </span></span>
<span><span class="co">#&gt; # dataNames(), dataParams(), dataNotes(), dataPaths(), dataTag() or mcols()</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;          name        Path                                  </span></span>
<span><span class="co">#&gt;   BFC1 | outfile.txt /tmp/RtmpXMx6sd/SharedData/outfile.txt</span></span>
<span><span class="co">## dh["BFC1"]</span></span>
<span><span class="va">dh</span><span class="op">[</span><span class="fu"><a href="../reference/dataHub-class.html">dataNames</a></span><span class="op">(</span><span class="va">dh</span><span class="op">)</span> <span class="op">==</span> <span class="st">"outfile.txt"</span><span class="op">]</span></span>
<span><span class="co">#&gt; dataHub with 1 records</span></span>
<span><span class="co">#&gt; cache path:  /tmp/RtmpXMx6sd/cache/ReUseData </span></span>
<span><span class="co">#&gt; # dataUpdate() to update the local data cache</span></span>
<span><span class="co">#&gt; # dataSearch() to query a specific dataset</span></span>
<span><span class="co">#&gt; # Additional information can be retrieved using: </span></span>
<span><span class="co">#&gt; # dataNames(), dataParams(), dataNotes(), dataPaths(), dataTag() or mcols()</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;          name        Path                                  </span></span>
<span><span class="co">#&gt;   BFC1 | outfile.txt /tmp/RtmpXMx6sd/SharedData/outfile.txt</span></span>
<span><span class="fu"><a href="../reference/dataHub-class.html">dataNames</a></span><span class="op">(</span><span class="va">dh</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "outfile.txt"</span></span>
<span><span class="fu"><a href="../reference/dataHub-class.html">dataParams</a></span><span class="op">(</span><span class="va">dh</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "input: Hello World!; outfile: outfile"</span></span>
<span><span class="fu"><a href="../reference/dataHub-class.html">dataNotes</a></span><span class="op">(</span><span class="va">dh</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "echo hello world txt"</span></span>
<span><span class="fu"><a href="../reference/dataHub-class.html">dataTags</a></span><span class="op">(</span><span class="va">dh</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] ""</span></span>
<span><span class="fu"><a href="../reference/dataHub-class.html">dataYml</a></span><span class="op">(</span><span class="va">dh</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "/tmp/RtmpXMx6sd/SharedData/echo_out_Hello_World!_outfile.yml"</span></span></code></pre></div>
<p><code>ReUseData</code>, as the name suggests, commits to promoting
the data reuse. Data can be prepared in standard input formats
(<code>toList</code>), e.g., YAML and JSON, to be easily integrated in
workflow methods that are locally or cloud-hosted.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">dh1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dataSearch.html">dataSearch</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"echo"</span>, <span class="st">"hello"</span>, <span class="st">"world"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; dataHub with 1 records</span></span>
<span><span class="co">#&gt; cache path:  /tmp/RtmpXMx6sd/cache/ReUseData </span></span>
<span><span class="co">#&gt; # dataUpdate() to update the local data cache</span></span>
<span><span class="co">#&gt; # dataSearch() to query a specific dataset</span></span>
<span><span class="co">#&gt; # Additional information can be retrieved using: </span></span>
<span><span class="co">#&gt; # dataNames(), dataParams(), dataNotes(), dataPaths(), dataTag() or mcols()</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;          name        Path                                  </span></span>
<span><span class="co">#&gt;   BFC1 | outfile.txt /tmp/RtmpXMx6sd/SharedData/outfile.txt</span></span>
<span><span class="fu"><a href="../reference/dataHub-class.html">toList</a></span><span class="op">(</span><span class="va">dh1</span>, listNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"input_file"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; $input_file</span></span>
<span><span class="co">#&gt; [1] "/tmp/RtmpXMx6sd/SharedData/outfile.txt"</span></span>
<span><span class="fu"><a href="../reference/dataHub-class.html">toList</a></span><span class="op">(</span><span class="va">dh1</span>, format <span class="op">=</span> <span class="st">"yaml"</span>, listNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"input_file"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "input_file: /tmp/RtmpXMx6sd/SharedData/outfile.txt"</span></span>
<span><span class="fu"><a href="../reference/dataHub-class.html">toList</a></span><span class="op">(</span><span class="va">dh1</span>, format <span class="op">=</span> <span class="st">"json"</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"data.json"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; File is saved as: "/tmp/RtmpXMx6sd/data.json"</span></span>
<span><span class="co">#&gt; {</span></span>
<span><span class="co">#&gt;   "outfile.txt": "/tmp/RtmpXMx6sd/SharedData/outfile.txt"</span></span>
<span><span class="co">#&gt; }</span></span></code></pre></div>
<p>Data can also be aggregated from different resources by tagging with
specific software tools.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/dataSearch.html">dataSearch</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; dataHub with 1 records</span></span>
<span><span class="co">#&gt; cache path:  /tmp/RtmpXMx6sd/cache/ReUseData </span></span>
<span><span class="co">#&gt; # dataUpdate() to update the local data cache</span></span>
<span><span class="co">#&gt; # dataSearch() to query a specific dataset</span></span>
<span><span class="co">#&gt; # Additional information can be retrieved using: </span></span>
<span><span class="co">#&gt; # dataNames(), dataParams(), dataNotes(), dataPaths(), dataTag() or mcols()</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;          name        Path                                  </span></span>
<span><span class="co">#&gt;   BFC1 | outfile.txt /tmp/RtmpXMx6sd/SharedData/outfile.txt</span></span>
<span><span class="fu"><a href="../reference/dataHub-class.html">dataTags</a></span><span class="op">(</span><span class="va">dh</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"#gatk"</span></span>
<span><span class="fu"><a href="../reference/dataSearch.html">dataSearch</a></span><span class="op">(</span><span class="st">"#gatk"</span><span class="op">)</span></span>
<span><span class="co">#&gt; dataHub with 1 records</span></span>
<span><span class="co">#&gt; cache path:  /tmp/RtmpXMx6sd/cache/ReUseData </span></span>
<span><span class="co">#&gt; # dataUpdate() to update the local data cache</span></span>
<span><span class="co">#&gt; # dataSearch() to query a specific dataset</span></span>
<span><span class="co">#&gt; # Additional information can be retrieved using: </span></span>
<span><span class="co">#&gt; # dataNames(), dataParams(), dataNotes(), dataPaths(), dataTag() or mcols()</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;          name        Path                                  </span></span>
<span><span class="co">#&gt;   BFC1 | outfile.txt /tmp/RtmpXMx6sd/SharedData/outfile.txt</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="existing-data-annotation">Existing data annotation<a class="anchor" aria-label="anchor" href="#existing-data-annotation"></a>
</h3>
<p>The package can also be used to add annotation and notes to existing
data resources or experiment data for management. Here we add exisiting
“exp_data” to local data repository.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exp_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"exp_data"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">exp_data</span><span class="op">)</span></span></code></pre></div>
<p>We first add notes to the data, and then update data repository with
information from the new dataset.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/annData.html">annData</a></span><span class="op">(</span><span class="va">exp_data</span>, notes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"experiment data"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; meta.yml added</span></span>
<span><span class="co">#&gt; [1] "/tmp/RtmpXMx6sd/exp_data/meta.yml"</span></span>
<span><span class="fu"><a href="../reference/dataUpdate.html">dataUpdate</a></span><span class="op">(</span><span class="va">exp_data</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Updating data record...</span></span>
<span><span class="co">#&gt; exp_data added</span></span>
<span><span class="co">#&gt; dataHub with 2 records</span></span>
<span><span class="co">#&gt; cache path:  /tmp/RtmpXMx6sd/cache/ReUseData </span></span>
<span><span class="co">#&gt; # dataUpdate() to update the local data cache</span></span>
<span><span class="co">#&gt; # dataSearch() to query a specific dataset</span></span>
<span><span class="co">#&gt; # Additional information can be retrieved using: </span></span>
<span><span class="co">#&gt; # dataNames(), dataParams(), dataNotes(), dataPaths(), dataTag() or mcols()</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;          name        Path                                  </span></span>
<span><span class="co">#&gt;   BFC1 | outfile.txt /tmp/RtmpXMx6sd/SharedData/outfile.txt</span></span>
<span><span class="co">#&gt;   BFC2 | exp_data    /tmp/RtmpXMx6sd/exp_data</span></span></code></pre></div>
<p>Now our data hub cached meta information from two different
directories, one from data recipe and one from exisiting data. Data can
be retrieved by keywords.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/dataSearch.html">dataSearch</a></span><span class="op">(</span><span class="st">"experiment"</span><span class="op">)</span></span>
<span><span class="co">#&gt; dataHub with 1 records</span></span>
<span><span class="co">#&gt; cache path:  /tmp/RtmpXMx6sd/cache/ReUseData </span></span>
<span><span class="co">#&gt; # dataUpdate() to update the local data cache</span></span>
<span><span class="co">#&gt; # dataSearch() to query a specific dataset</span></span>
<span><span class="co">#&gt; # Additional information can be retrieved using: </span></span>
<span><span class="co">#&gt; # dataNames(), dataParams(), dataNotes(), dataPaths(), dataTag() or mcols()</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;          name     Path                    </span></span>
<span><span class="co">#&gt;   BFC2 | exp_data /tmp/RtmpXMx6sd/exp_data</span></span></code></pre></div>
<p><strong>NOTE:</strong> if the argument <code>cloud=TRUE</code> is
enabled, <code><a href="../reference/dataUpdate.html">dataUpdate()</a></code> will also cache the pregenerated data
sets (from evaluation of public ReUseData recipes) that are available on
ReUseData google bucket and return in the <code>dataHub</code> object
that are fully searchable. Please see the following section for
details.</p>
</div>
</div>
<div class="section level2">
<h2 id="cloud-data-resources">Cloud data resources<a class="anchor" aria-label="anchor" href="#cloud-data-resources"></a>
</h2>
<p>With the prebuilt data recipes for curation (e.g., downloading,
unzipping, indexing) of commonly used public data resources we have
pregenerated some data sets and put them on the cloud space for direct
use.</p>
<p>Before searching, one need to use <code>dataUpdate(cloud=TRUE)</code>
to sync the existing data sets on cloud, then <code><a href="../reference/dataSearch.html">dataSearch()</a></code>
can be used to search any available data set either in local cache and
on the cloud.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gcpdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"gcpData"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/dataUpdate.html">dataUpdate</a></span><span class="op">(</span><span class="va">gcpdir</span>, cloud<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Updating data record...</span></span>
<span><span class="co">#&gt; 35ddd53464774_GRCh38.primary_assembly.genome.fa.1.bt2 added</span></span>
<span><span class="co">#&gt; 35ddd24d1ee6f_GRCh38.primary_assembly.genome.fa.2.bt2 added</span></span>
<span><span class="co">#&gt; 35ddd55648fe3_GRCh38.primary_assembly.genome.fa.3.bt2 added</span></span>
<span><span class="co">#&gt; 35ddd2a852212_GRCh38.primary_assembly.genome.fa.4.bt2 added</span></span>
<span><span class="co">#&gt; 35ddd499a6fda_GRCh38.primary_assembly.genome.fa.rev.1.bt2 added</span></span>
<span><span class="co">#&gt; 35ddd5cbe42ee_GRCh38.primary_assembly.genome.fa.rev.2.bt2 added</span></span>
<span><span class="co">#&gt; 35ddd2598047d_outfile.txt added</span></span>
<span><span class="co">#&gt; 35ddd2d330d9_GRCh37_to_GRCh38.chain added</span></span>
<span><span class="co">#&gt; 35ddd5a31bf7_GRCh37_to_NCBI34.chain added</span></span>
<span><span class="co">#&gt; 35ddd49f79d5d_GRCh37_to_NCBI35.chain added</span></span>
<span><span class="co">#&gt; 35dddf75980e_GRCh37_to_NCBI36.chain added</span></span>
<span><span class="co">#&gt; 35ddd2d15ce58_GRCh38_to_GRCh37.chain added</span></span>
<span><span class="co">#&gt; 35ddd6eca4166_GRCh38_to_NCBI34.chain added</span></span>
<span><span class="co">#&gt; 35ddd4f0e372a_GRCh38_to_NCBI35.chain added</span></span>
<span><span class="co">#&gt; 35ddd27a79a04_GRCh38_to_NCBI36.chain added</span></span>
<span><span class="co">#&gt; 35ddd7696527d_NCBI34_to_GRCh37.chain added</span></span>
<span><span class="co">#&gt; 35dddf8757b0_NCBI34_to_GRCh38.chain added</span></span>
<span><span class="co">#&gt; 35ddd658bb06e_NCBI35_to_GRCh37.chain added</span></span>
<span><span class="co">#&gt; 35ddd663bca71_NCBI35_to_GRCh38.chain added</span></span>
<span><span class="co">#&gt; 35ddd3abdd4f_NCBI36_to_GRCh37.chain added</span></span>
<span><span class="co">#&gt; 35ddd3eb5025b_NCBI36_to_GRCh38.chain added</span></span>
<span><span class="co">#&gt; 35ddd12c34cb1_GRCm38_to_NCBIM36.chain added</span></span>
<span><span class="co">#&gt; 35ddd521d52d4_GRCm38_to_NCBIM37.chain added</span></span>
<span><span class="co">#&gt; 35ddd457bb785_NCBIM36_to_GRCm38.chain added</span></span>
<span><span class="co">#&gt; 35ddd68fff932_NCBIM37_to_GRCm38.chain added</span></span>
<span><span class="co">#&gt; 35dddf63b2e6_1000G_omni2.5.b37.vcf.gz added</span></span>
<span><span class="co">#&gt; 35ddd6055da0c_1000G_omni2.5.b37.vcf.gz.tbi added</span></span>
<span><span class="co">#&gt; 35ddd1ab67c12_Mills_and_1000G_gold_standard.indels.b37.vcf.gz added</span></span>
<span><span class="co">#&gt; 35ddd32f81c4b_Mills_and_1000G_gold_standard.indels.b37.vcf.gz.tbi added</span></span>
<span><span class="co">#&gt; 35ddd163ef277_1000G_omni2.5.hg38.vcf.gz added</span></span>
<span><span class="co">#&gt; 35ddd25d4cc50_1000G_omni2.5.hg38.vcf.gz.tbi added</span></span>
<span><span class="co">#&gt; 35ddd63e63bf_Mills_and_1000G_gold_standard.indels.hg38.vcf.gz added</span></span>
<span><span class="co">#&gt; 35ddd3b10e0e6_Mills_and_1000G_gold_standard.indels.hg38.vcf.gz.tbi added</span></span>
<span><span class="co">#&gt; 35ddd7b395c33_af-only-gnomad.raw.sites.vcf added</span></span>
<span><span class="co">#&gt; 35ddd30c385d1_af-only-gnomad.raw.sites.vcf.idx added</span></span>
<span><span class="co">#&gt; 35ddd4ab50c1_Mutect2-exome-panel.vcf.idx added</span></span>
<span><span class="co">#&gt; 35ddd57f79f21_Mutect2-WGS-panel-b37.vcf added</span></span>
<span><span class="co">#&gt; 35ddd565b8a4e_Mutect2-WGS-panel-b37.vcf.idx added</span></span>
<span><span class="co">#&gt; 35ddd77e819a_small_exac_common_3.vcf added</span></span>
<span><span class="co">#&gt; 35ddd5d9abb19_small_exac_common_3.vcf.idx added</span></span>
<span><span class="co">#&gt; 35ddd205327ab_1000g_pon.hg38.vcf.gz added</span></span>
<span><span class="co">#&gt; 35ddd16f419a8_1000g_pon.hg38.vcf.gz.tbi added</span></span>
<span><span class="co">#&gt; 35dddab08971_af-only-gnomad.hg38.vcf.gz added</span></span>
<span><span class="co">#&gt; 35dddf1d6911_af-only-gnomad.hg38.vcf.gz.tbi added</span></span>
<span><span class="co">#&gt; 35ddd660250d2_small_exac_common_3.hg38.vcf.gz added</span></span>
<span><span class="co">#&gt; 35ddd32582375_small_exac_common_3.hg38.vcf.gz.tbi added</span></span>
<span><span class="co">#&gt; 35ddd5b3bb8e_gencode.v41.annotation.gtf added</span></span>
<span><span class="co">#&gt; 35ddd7589a882_gencode.v42.annotation.gtf added</span></span>
<span><span class="co">#&gt; 35ddd17e3d3e4_gencode.vM30.annotation.gtf added</span></span>
<span><span class="co">#&gt; 35ddd6bef8600_gencode.vM31.annotation.gtf added</span></span>
<span><span class="co">#&gt; 35ddd793585d1_gencode.v41.transcripts.fa added</span></span>
<span><span class="co">#&gt; 35ddd5698d63f_gencode.v41.transcripts.fa.fai added</span></span>
<span><span class="co">#&gt; 35ddd7eb2d2b1_gencode.v42.transcripts.fa added</span></span>
<span><span class="co">#&gt; 35ddd4b52d8a5_gencode.v42.transcripts.fa.fai added</span></span>
<span><span class="co">#&gt; 35ddd1c148dc4_gencode.vM30.pc_transcripts.fa added</span></span>
<span><span class="co">#&gt; 35ddd67b2cbe3_gencode.vM30.pc_transcripts.fa.fai added</span></span>
<span><span class="co">#&gt; 35ddd5ab68b8c_gencode.vM31.pc_transcripts.fa added</span></span>
<span><span class="co">#&gt; 35ddd7c6a67d0_gencode.vM31.pc_transcripts.fa.fai added</span></span>
<span><span class="co">#&gt; 35ddd26947f6_GRCh38.primary_assembly.genome.fa.1.ht2 added</span></span>
<span><span class="co">#&gt; 35ddddaea7d7_GRCh38.primary_assembly.genome.fa.2.ht2 added</span></span>
<span><span class="co">#&gt; 35ddd12a95a48_GRCh38.primary_assembly.genome.fa.3.ht2 added</span></span>
<span><span class="co">#&gt; 35ddd283e1446_GRCh38.primary_assembly.genome.fa.4.ht2 added</span></span>
<span><span class="co">#&gt; 35ddd13ed0b96_GRCh38.primary_assembly.genome.fa.5.ht2 added</span></span>
<span><span class="co">#&gt; 35ddd4dba3b2e_GRCh38.primary_assembly.genome.fa.6.ht2 added</span></span>
<span><span class="co">#&gt; 35ddd23777079_GRCh38.primary_assembly.genome.fa.7.ht2 added</span></span>
<span><span class="co">#&gt; 35ddd44b09167_GRCh38.primary_assembly.genome.fa.8.ht2 added</span></span>
<span><span class="co">#&gt; 35ddd52658bef_GRCh38_full_analysis_set_plus_decoy_hla.fa.fai added</span></span>
<span><span class="co">#&gt; 35ddd7b6f0f9b_GRCh38_full_analysis_set_plus_decoy_hla.fa.amb added</span></span>
<span><span class="co">#&gt; 35ddd1b0c1bb5_GRCh38_full_analysis_set_plus_decoy_hla.fa.ann added</span></span>
<span><span class="co">#&gt; 35ddd59e40d89_GRCh38_full_analysis_set_plus_decoy_hla.fa.bwt added</span></span>
<span><span class="co">#&gt; 35ddd5909cab4_GRCh38_full_analysis_set_plus_decoy_hla.fa.pac added</span></span>
<span><span class="co">#&gt; 35ddd3b5f4361_GRCh38_full_analysis_set_plus_decoy_hla.fa.sa added</span></span>
<span><span class="co">#&gt; 35ddd70d82731_GRCh38_full_analysis_set_plus_decoy_hla.fa added</span></span>
<span><span class="co">#&gt; 35ddd63ba5425_GRCh38.primary_assembly.genome.fa.fai added</span></span>
<span><span class="co">#&gt; 35ddd4a7cac72_GRCh38.primary_assembly.genome.fa.amb added</span></span>
<span><span class="co">#&gt; 35ddd56da7803_GRCh38.primary_assembly.genome.fa.ann added</span></span>
<span><span class="co">#&gt; 35ddd1612779b_GRCh38.primary_assembly.genome.fa.bwt added</span></span>
<span><span class="co">#&gt; 35ddd50306801_GRCh38.primary_assembly.genome.fa.pac added</span></span>
<span><span class="co">#&gt; 35ddd4c642085_GRCh38.primary_assembly.genome.fa.sa added</span></span>
<span><span class="co">#&gt; 35ddd2df64b7f_GRCh38.primary_assembly.genome.fa added</span></span>
<span><span class="co">#&gt; 35ddd3c1fee01_hs37d5.fa.fai added</span></span>
<span><span class="co">#&gt; 35ddd4599a657_hs37d5.fa.amb added</span></span>
<span><span class="co">#&gt; 35ddd48f21be_hs37d5.fa.ann added</span></span>
<span><span class="co">#&gt; 35ddd3ad2c0b2_hs37d5.fa.bwt added</span></span>
<span><span class="co">#&gt; 35ddd10ec7efc_hs37d5.fa.pac added</span></span>
<span><span class="co">#&gt; 35ddd20a3af83_hs37d5.fa.sa added</span></span>
<span><span class="co">#&gt; 35ddd22858c96_hs37d5.fa added</span></span>
<span><span class="co">#&gt; 35ddd6ba30a88_complete_ref_lens.bin added</span></span>
<span><span class="co">#&gt; 35ddd1d0e1753_ctable.bin added</span></span>
<span><span class="co">#&gt; 35ddd24eed48c_ctg_offsets.bin added</span></span>
<span><span class="co">#&gt; 35ddd7951b25f_duplicate_clusters.tsv added</span></span>
<span><span class="co">#&gt; 35ddd2fb7719b_info.json added</span></span>
<span><span class="co">#&gt; 35ddd4d2ce8d2_mphf.bin added</span></span>
<span><span class="co">#&gt; 35dddd3ebdf5_pos.bin added</span></span>
<span><span class="co">#&gt; 35ddd7d71acca_pre_indexing.log added</span></span>
<span><span class="co">#&gt; 35ddd70a4594b_rank.bin added</span></span>
<span><span class="co">#&gt; 35ddd51ef4f5c_ref_indexing.log added</span></span>
<span><span class="co">#&gt; 35ddd4fd738b9_refAccumLengths.bin added</span></span>
<span><span class="co">#&gt; 35ddd6c1368e6_reflengths.bin added</span></span>
<span><span class="co">#&gt; 35ddd6cfb6b12_refseq.bin added</span></span>
<span><span class="co">#&gt; 35ddd29bb4643_seq.bin added</span></span>
<span><span class="co">#&gt; 35ddd451d339a_versionInfo.json added</span></span>
<span><span class="co">#&gt; 35ddd285aae73_salmon_index added</span></span>
<span><span class="co">#&gt; 35ddd1a936d74_chrLength.txt added</span></span>
<span><span class="co">#&gt; 35ddd28d787c0_chrName.txt added</span></span>
<span><span class="co">#&gt; 35ddd72d75ae5_chrNameLength.txt added</span></span>
<span><span class="co">#&gt; 35ddd716de578_chrStart.txt added</span></span>
<span><span class="co">#&gt; 35ddd3ee9ff5b_exonGeTrInfo.tab added</span></span>
<span><span class="co">#&gt; 35ddd4307c2e6_exonInfo.tab added</span></span>
<span><span class="co">#&gt; 35ddd3dd205fd_geneInfo.tab added</span></span>
<span><span class="co">#&gt; 35ddd6ce04ada_Genome added</span></span>
<span><span class="co">#&gt; 35ddd7f27b0e7_genomeParameters.txt added</span></span>
<span><span class="co">#&gt; 35ddd36bac54_Log.out added</span></span>
<span><span class="co">#&gt; 35ddd716f6c98_SA added</span></span>
<span><span class="co">#&gt; 35ddd39fa719a_SAindex added</span></span>
<span><span class="co">#&gt; 35ddd14582b51_sjdbInfo.txt added</span></span>
<span><span class="co">#&gt; 35ddd12131c1b_sjdbList.fromGTF.out.tab added</span></span>
<span><span class="co">#&gt; 35ddd5c7ffe30_sjdbList.out.tab added</span></span>
<span><span class="co">#&gt; 35ddd7ffb35d9_transcriptInfo.tab added</span></span>
<span><span class="co">#&gt; 35ddd2f21336f_GRCh38.GENCODE.v42_100 added</span></span>
<span><span class="co">#&gt; 35ddd16ed2bc_knownGene_hg38.sql added</span></span>
<span><span class="co">#&gt; 35ddd794ce839_knownGene_hg38.txt added</span></span>
<span><span class="co">#&gt; 35ddd5ed8a50a_refGene_hg38.sql added</span></span>
<span><span class="co">#&gt; 35ddd4e9bbb8e_refGene_hg38.txt added</span></span>
<span><span class="co">#&gt; 35ddd68ba62e_knownGene_mm39.sql added</span></span>
<span><span class="co">#&gt; 35ddd5c4a51d4_knownGene_mm39.txt added</span></span>
<span><span class="co">#&gt; 35ddd3f4014d9_refGene_mm39.sql added</span></span>
<span><span class="co">#&gt; 35ddd587af58b_refGene_mm39.txt added</span></span>
<span><span class="co">#&gt; dataHub with 130 records</span></span>
<span><span class="co">#&gt; cache path:  /tmp/RtmpXMx6sd/cache/ReUseData </span></span>
<span><span class="co">#&gt; # dataUpdate() to update the local data cache</span></span>
<span><span class="co">#&gt; # dataSearch() to query a specific dataset</span></span>
<span><span class="co">#&gt; # Additional information can be retrieved using: </span></span>
<span><span class="co">#&gt; # dataNames(), dataParams(), dataNotes(), dataPaths(), dataTag() or mcols()</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;            name                                   </span></span>
<span><span class="co">#&gt;   BFC1   | outfile.txt                            </span></span>
<span><span class="co">#&gt;   BFC2   | exp_data                               </span></span>
<span><span class="co">#&gt;   BFC3   | GRCh38.primary_assembly.genome.fa.1.bt2</span></span>
<span><span class="co">#&gt;   BFC4   | GRCh38.primary_assembly.genome.fa.2.bt2</span></span>
<span><span class="co">#&gt;   BFC5   | GRCh38.primary_assembly.genome.fa.3.bt2</span></span>
<span><span class="co">#&gt;   ...      ...                                    </span></span>
<span><span class="co">#&gt;   BFC126 | refGene_hg38.txt                       </span></span>
<span><span class="co">#&gt;   BFC127 | knownGene_mm39.sql                     </span></span>
<span><span class="co">#&gt;   BFC128 | knownGene_mm39.txt                     </span></span>
<span><span class="co">#&gt;   BFC129 | refGene_mm39.sql                       </span></span>
<span><span class="co">#&gt;   BFC130 | refGene_mm39.txt                       </span></span>
<span><span class="co">#&gt;          Path                                                               </span></span>
<span><span class="co">#&gt;   BFC1   /tmp/RtmpXMx6sd/SharedData/outfile.txt                             </span></span>
<span><span class="co">#&gt;   BFC2   /tmp/RtmpXMx6sd/exp_data                                           </span></span>
<span><span class="co">#&gt;   BFC3   https://storage.googleapis.com/reusedata/bowtie2_index/GRCh38.pr...</span></span>
<span><span class="co">#&gt;   BFC4   https://storage.googleapis.com/reusedata/bowtie2_index/GRCh38.pr...</span></span>
<span><span class="co">#&gt;   BFC5   https://storage.googleapis.com/reusedata/bowtie2_index/GRCh38.pr...</span></span>
<span><span class="co">#&gt;   ...    ...                                                                </span></span>
<span><span class="co">#&gt;   BFC126 https://storage.googleapis.com/reusedata/ucsc_database/refGene_h...</span></span>
<span><span class="co">#&gt;   BFC127 https://storage.googleapis.com/reusedata/ucsc_database/knownGene...</span></span>
<span><span class="co">#&gt;   BFC128 https://storage.googleapis.com/reusedata/ucsc_database/knownGene...</span></span>
<span><span class="co">#&gt;   BFC129 https://storage.googleapis.com/reusedata/ucsc_database/refGene_m...</span></span>
<span><span class="co">#&gt;   BFC130 https://storage.googleapis.com/reusedata/ucsc_database/refGene_m...</span></span></code></pre></div>
<p>If the data of interest already exist on the cloud, then
<code>getCloudData</code> will directly download the data to your
computer. Add it to the local caching system using
<code><a href="../reference/dataUpdate.html">dataUpdate()</a></code> for later use.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">dh</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dataSearch.html">dataSearch</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ensembl"</span>, <span class="st">"GRCh38"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; dataHub with 8 records</span></span>
<span><span class="co">#&gt; cache path:  /tmp/RtmpXMx6sd/cache/ReUseData </span></span>
<span><span class="co">#&gt; # dataUpdate() to update the local data cache</span></span>
<span><span class="co">#&gt; # dataSearch() to query a specific dataset</span></span>
<span><span class="co">#&gt; # Additional information can be retrieved using: </span></span>
<span><span class="co">#&gt; # dataNames(), dataParams(), dataNotes(), dataPaths(), dataTag() or mcols()</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;           name                  </span></span>
<span><span class="co">#&gt;   BFC10 | GRCh37_to_GRCh38.chain</span></span>
<span><span class="co">#&gt;   BFC14 | GRCh38_to_GRCh37.chain</span></span>
<span><span class="co">#&gt;   BFC15 | GRCh38_to_NCBI34.chain</span></span>
<span><span class="co">#&gt;   BFC16 | GRCh38_to_NCBI35.chain</span></span>
<span><span class="co">#&gt;   BFC17 | GRCh38_to_NCBI36.chain</span></span>
<span><span class="co">#&gt;   BFC19 | NCBI34_to_GRCh38.chain</span></span>
<span><span class="co">#&gt;   BFC21 | NCBI35_to_GRCh38.chain</span></span>
<span><span class="co">#&gt;   BFC23 | NCBI36_to_GRCh38.chain</span></span>
<span><span class="co">#&gt;         Path                                                               </span></span>
<span><span class="co">#&gt;   BFC10 https://storage.googleapis.com/reusedata/ensembl_liftover/GRCh37...</span></span>
<span><span class="co">#&gt;   BFC14 https://storage.googleapis.com/reusedata/ensembl_liftover/GRCh38...</span></span>
<span><span class="co">#&gt;   BFC15 https://storage.googleapis.com/reusedata/ensembl_liftover/GRCh38...</span></span>
<span><span class="co">#&gt;   BFC16 https://storage.googleapis.com/reusedata/ensembl_liftover/GRCh38...</span></span>
<span><span class="co">#&gt;   BFC17 https://storage.googleapis.com/reusedata/ensembl_liftover/GRCh38...</span></span>
<span><span class="co">#&gt;   BFC19 https://storage.googleapis.com/reusedata/ensembl_liftover/NCBI34...</span></span>
<span><span class="co">#&gt;   BFC21 https://storage.googleapis.com/reusedata/ensembl_liftover/NCBI35...</span></span>
<span><span class="co">#&gt;   BFC23 https://storage.googleapis.com/reusedata/ensembl_liftover/NCBI36...</span></span>
<span><span class="fu"><a href="../reference/getCloudData.html">getCloudData</a></span><span class="op">(</span><span class="va">dh</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, outdir <span class="op">=</span> <span class="va">gcpdir</span><span class="op">)</span></span>
<span><span class="co">#&gt; Data is downloaded: </span></span>
<span><span class="co">#&gt; /tmp/RtmpXMx6sd/gcpData/GRCh37_to_GRCh38.chain</span></span></code></pre></div>
<p>Now we create the data cache with only local data files, and we can
see that the downloaded data is available.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/dataUpdate.html">dataUpdate</a></span><span class="op">(</span><span class="va">gcpdir</span><span class="op">)</span>  <span class="co">## Update local data cache (without cloud data)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Updating data record...</span></span>
<span><span class="co">#&gt; GRCh37_to_GRCh38.chain added</span></span>
<span><span class="co">#&gt; dataHub with 131 records</span></span>
<span><span class="co">#&gt; cache path:  /tmp/RtmpXMx6sd/cache/ReUseData </span></span>
<span><span class="co">#&gt; # dataUpdate() to update the local data cache</span></span>
<span><span class="co">#&gt; # dataSearch() to query a specific dataset</span></span>
<span><span class="co">#&gt; # Additional information can be retrieved using: </span></span>
<span><span class="co">#&gt; # dataNames(), dataParams(), dataNotes(), dataPaths(), dataTag() or mcols()</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;            name                                   </span></span>
<span><span class="co">#&gt;   BFC1   | outfile.txt                            </span></span>
<span><span class="co">#&gt;   BFC2   | exp_data                               </span></span>
<span><span class="co">#&gt;   BFC3   | GRCh38.primary_assembly.genome.fa.1.bt2</span></span>
<span><span class="co">#&gt;   BFC4   | GRCh38.primary_assembly.genome.fa.2.bt2</span></span>
<span><span class="co">#&gt;   BFC5   | GRCh38.primary_assembly.genome.fa.3.bt2</span></span>
<span><span class="co">#&gt;   ...      ...                                    </span></span>
<span><span class="co">#&gt;   BFC127 | knownGene_mm39.sql                     </span></span>
<span><span class="co">#&gt;   BFC128 | knownGene_mm39.txt                     </span></span>
<span><span class="co">#&gt;   BFC129 | refGene_mm39.sql                       </span></span>
<span><span class="co">#&gt;   BFC130 | refGene_mm39.txt                       </span></span>
<span><span class="co">#&gt;   BFC131 | GRCh37_to_GRCh38.chain                 </span></span>
<span><span class="co">#&gt;          Path                                                               </span></span>
<span><span class="co">#&gt;   BFC1   /tmp/RtmpXMx6sd/SharedData/outfile.txt                             </span></span>
<span><span class="co">#&gt;   BFC2   /tmp/RtmpXMx6sd/exp_data                                           </span></span>
<span><span class="co">#&gt;   BFC3   https://storage.googleapis.com/reusedata/bowtie2_index/GRCh38.pr...</span></span>
<span><span class="co">#&gt;   BFC4   https://storage.googleapis.com/reusedata/bowtie2_index/GRCh38.pr...</span></span>
<span><span class="co">#&gt;   BFC5   https://storage.googleapis.com/reusedata/bowtie2_index/GRCh38.pr...</span></span>
<span><span class="co">#&gt;   ...    ...                                                                </span></span>
<span><span class="co">#&gt;   BFC127 https://storage.googleapis.com/reusedata/ucsc_database/knownGene...</span></span>
<span><span class="co">#&gt;   BFC128 https://storage.googleapis.com/reusedata/ucsc_database/knownGene...</span></span>
<span><span class="co">#&gt;   BFC129 https://storage.googleapis.com/reusedata/ucsc_database/refGene_m...</span></span>
<span><span class="co">#&gt;   BFC130 https://storage.googleapis.com/reusedata/ucsc_database/refGene_m...</span></span>
<span><span class="co">#&gt;   BFC131 /tmp/RtmpXMx6sd/gcpData/GRCh37_to_GRCh38.chain</span></span>
<span><span class="fu"><a href="../reference/dataSearch.html">dataSearch</a></span><span class="op">(</span><span class="op">)</span>  <span class="co">## data is available locally!!!</span></span>
<span><span class="co">#&gt; dataHub with 131 records</span></span>
<span><span class="co">#&gt; cache path:  /tmp/RtmpXMx6sd/cache/ReUseData </span></span>
<span><span class="co">#&gt; # dataUpdate() to update the local data cache</span></span>
<span><span class="co">#&gt; # dataSearch() to query a specific dataset</span></span>
<span><span class="co">#&gt; # Additional information can be retrieved using: </span></span>
<span><span class="co">#&gt; # dataNames(), dataParams(), dataNotes(), dataPaths(), dataTag() or mcols()</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;            name                                   </span></span>
<span><span class="co">#&gt;   BFC1   | outfile.txt                            </span></span>
<span><span class="co">#&gt;   BFC2   | exp_data                               </span></span>
<span><span class="co">#&gt;   BFC3   | GRCh38.primary_assembly.genome.fa.1.bt2</span></span>
<span><span class="co">#&gt;   BFC4   | GRCh38.primary_assembly.genome.fa.2.bt2</span></span>
<span><span class="co">#&gt;   BFC5   | GRCh38.primary_assembly.genome.fa.3.bt2</span></span>
<span><span class="co">#&gt;   ...      ...                                    </span></span>
<span><span class="co">#&gt;   BFC127 | knownGene_mm39.sql                     </span></span>
<span><span class="co">#&gt;   BFC128 | knownGene_mm39.txt                     </span></span>
<span><span class="co">#&gt;   BFC129 | refGene_mm39.sql                       </span></span>
<span><span class="co">#&gt;   BFC130 | refGene_mm39.txt                       </span></span>
<span><span class="co">#&gt;   BFC131 | GRCh37_to_GRCh38.chain                 </span></span>
<span><span class="co">#&gt;          Path                                                               </span></span>
<span><span class="co">#&gt;   BFC1   /tmp/RtmpXMx6sd/SharedData/outfile.txt                             </span></span>
<span><span class="co">#&gt;   BFC2   /tmp/RtmpXMx6sd/exp_data                                           </span></span>
<span><span class="co">#&gt;   BFC3   https://storage.googleapis.com/reusedata/bowtie2_index/GRCh38.pr...</span></span>
<span><span class="co">#&gt;   BFC4   https://storage.googleapis.com/reusedata/bowtie2_index/GRCh38.pr...</span></span>
<span><span class="co">#&gt;   BFC5   https://storage.googleapis.com/reusedata/bowtie2_index/GRCh38.pr...</span></span>
<span><span class="co">#&gt;   ...    ...                                                                </span></span>
<span><span class="co">#&gt;   BFC127 https://storage.googleapis.com/reusedata/ucsc_database/knownGene...</span></span>
<span><span class="co">#&gt;   BFC128 https://storage.googleapis.com/reusedata/ucsc_database/knownGene...</span></span>
<span><span class="co">#&gt;   BFC129 https://storage.googleapis.com/reusedata/ucsc_database/refGene_m...</span></span>
<span><span class="co">#&gt;   BFC130 https://storage.googleapis.com/reusedata/ucsc_database/refGene_m...</span></span>
<span><span class="co">#&gt;   BFC131 /tmp/RtmpXMx6sd/gcpData/GRCh37_to_GRCh38.chain</span></span></code></pre></div>
<p>The data supports user-friendly discovery and access through the
<code>ReUseData</code> portal, where detailed instructions are provided
for straight-forward incorporation into data analysis pipelines run on
local computing nodes, web resources, and cloud computing platforms
(e.g., Terra, CGC).</p>
</div>
<div class="section level2">
<h2 id="know-your-data">Know your data<a class="anchor" aria-label="anchor" href="#know-your-data"></a>
</h2>
<p>Here we provide a function <code><a href="../reference/meta_data.html">meta_data()</a></code> to create a data
frame that contains all information about the data sets in the specified
file path (recursively), including the annotation file
(<code>$yml</code> column), parameter values for the recipe
(<code>$params</code> column), data file path (<code>$output</code>
column), keywords for data file (<code>notes</code> columns), date of
data generation (<code>date</code> column), and any tag if available
(<code>tag</code> column).</p>
<p>Use <code>cleanup = TRUE</code> to cleanup any invalid or
expired/older intermediate files.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/meta_data.html">meta_data</a></span><span class="op">(</span><span class="va">outdir</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mt</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                                            yml</span></span>
<span><span class="co">#&gt; 1 /tmp/RtmpXMx6sd/SharedData/echo_out_Hello_World!_outfile.yml</span></span>
<span><span class="co">#&gt;                                  params                                 output</span></span>
<span><span class="co">#&gt; 1 input: Hello World!; outfile: outfile /tmp/RtmpXMx6sd/SharedData/outfile.txt</span></span>
<span><span class="co">#&gt;                  notes                       date</span></span>
<span><span class="co">#&gt; 1 echo hello world txt 2023-11-22 22:21:02.021824</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="sessioninfo">SessionInfo<a class="anchor" aria-label="anchor" href="#sessioninfo"></a>
</h2>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">#&gt; Platform: x86_64-conda-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS/LAPACK: /home/qhu/miniconda3/envs/rcwl/lib/libopenblasp-r0.3.23.so;  LAPACK version 3.11.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: America/New_York</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] ReUseData_1.3.2     Rcwl_1.18.0         S4Vectors_0.40.1   </span></span>
<span><span class="co">#&gt; [4] BiocGenerics_0.48.1 yaml_2.3.7          BiocStyle_2.30.0   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] tidyselect_1.2.0      dplyr_1.1.4           blob_1.2.4           </span></span>
<span><span class="co">#&gt;  [4] filelock_1.0.2        R.utils_2.12.3        fastmap_1.1.1        </span></span>
<span><span class="co">#&gt;  [7] BiocFileCache_2.10.1  promises_1.2.1        digest_0.6.33        </span></span>
<span><span class="co">#&gt; [10] base64url_1.4         mime_0.12             lifecycle_1.0.4      </span></span>
<span><span class="co">#&gt; [13] ellipsis_0.3.2        RSQLite_2.3.3         magrittr_2.0.3       </span></span>
<span><span class="co">#&gt; [16] compiler_4.3.1        rlang_1.1.2           sass_0.4.7           </span></span>
<span><span class="co">#&gt; [19] progress_1.2.2        tools_4.3.1           utf8_1.2.4           </span></span>
<span><span class="co">#&gt; [22] data.table_1.14.8     knitr_1.45            prettyunits_1.2.0    </span></span>
<span><span class="co">#&gt; [25] brew_1.0-8            htmlwidgets_1.6.2     curl_5.1.0           </span></span>
<span><span class="co">#&gt; [28] bit_4.0.5             reticulate_1.34.0     RColorBrewer_1.1-3   </span></span>
<span><span class="co">#&gt; [31] batchtools_0.9.17     BiocParallel_1.36.0   withr_2.5.2          </span></span>
<span><span class="co">#&gt; [34] purrr_1.0.2           desc_1.4.2            R.oo_1.25.0          </span></span>
<span><span class="co">#&gt; [37] grid_4.3.1            fansi_1.0.5           git2r_0.32.0         </span></span>
<span><span class="co">#&gt; [40] xtable_1.8-4          cli_3.6.1             rmarkdown_2.25       </span></span>
<span><span class="co">#&gt; [43] DiagrammeR_1.0.10     crayon_1.5.2          ragg_1.2.5           </span></span>
<span><span class="co">#&gt; [46] generics_0.1.3        httr_1.4.7            visNetwork_2.1.2     </span></span>
<span><span class="co">#&gt; [49] DBI_1.1.3             cachem_1.0.8          stringr_1.5.1        </span></span>
<span><span class="co">#&gt; [52] parallel_4.3.1        BiocManager_1.30.22   basilisk_1.14.1      </span></span>
<span><span class="co">#&gt; [55] vctrs_0.6.4           Matrix_1.6-3          jsonlite_1.8.7       </span></span>
<span><span class="co">#&gt; [58] dir.expiry_1.10.0     bookdown_0.36         hms_1.1.3            </span></span>
<span><span class="co">#&gt; [61] bit64_4.0.5           systemfonts_1.0.4     jquerylib_0.1.4      </span></span>
<span><span class="co">#&gt; [64] RcwlPipelines_1.18.0  glue_1.6.2            pkgdown_2.0.7        </span></span>
<span><span class="co">#&gt; [67] codetools_0.2-19      stringi_1.8.1         later_1.3.1          </span></span>
<span><span class="co">#&gt; [70] tibble_3.2.1          pillar_1.9.0          basilisk.utils_1.14.1</span></span>
<span><span class="co">#&gt; [73] rappdirs_0.3.3        htmltools_0.5.7       R6_2.5.1             </span></span>
<span><span class="co">#&gt; [76] dbplyr_2.4.0          textshaping_0.3.6     rprojroot_2.0.4      </span></span>
<span><span class="co">#&gt; [79] evaluate_0.23         shiny_1.8.0           lattice_0.22-5       </span></span>
<span><span class="co">#&gt; [82] R.methodsS3_1.8.2     png_0.1-8             backports_1.4.1      </span></span>
<span><span class="co">#&gt; [85] memoise_2.0.1         httpuv_1.6.12         bslib_0.5.1          </span></span>
<span><span class="co">#&gt; [88] Rcpp_1.0.11           checkmate_2.3.0       xfun_0.41            </span></span>
<span><span class="co">#&gt; [91] fs_1.6.3              pkgconfig_2.0.3</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Qian Liu.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
